cep-0016-2.md

```diff
@@ -0,0 +1,89 @@
+<table>
+<tr><td> Title </td><td> Run-exports in sharded Repodata. </td>
+<tr><td> Status </td><td> Draft </td></tr>
+<tr><td> Author(s) </td><td> Bas Zalmstra &lt;bas@prefix.dev&gt;</td></tr>
+<tr><td> Created </td><td> Jan 16, 2025</td></tr>
+<tr><td> Updated </td><td> Jan 16, 2025</td></tr>
+<tr><td> Discussion </td><td> </td></tr>
+<tr><td> Implementation </td><td>  </td></tr>
+</table>
+
+# Run-exports in sharded Repodata
+
+We propose to add run-export information to the sharded repodata shards.
+
+## Motivation
+
+When building conda packages the build infrastructure needs to extract run-export information from conda packages in the host- and build environments.
+Run-export information is stored in a package and can be extracted by downloading the package and extracting the `run_exports.json` file. 
+Even with the possibility to stream parts of `.conda` files this is a relatively resource-intensive operation.
+
+[CEP-12](https://github.com/conda/ceps/blob/main/cep-0012.md) formalized a `run_exports.json` file that is stored next `repodata.json` file. 
+However, not all channels on the default server (conda.anaconda.org) provide this information which means falling back to downloading and extracting this information from the packages. It is possible to extract the data by only sparsly reading the file but the overhead is still relatively large.
+
+Having two separate files also poses some problems as extra mechanisms have to be introduced in the build infrastructure to manage and sync both files on the build machines.
+
+## Specification
+
+CEP-12 mentions the following reasons for splitting the information into two files:
+
+> * It would require extending the repodata schema, currently not formally standardized.
+> * It would increase the size of the already heavy repodata.json files.
+> * (Typed) repodata parsers would need to be updated to handle the new field.
+We propose that these reasons no longer hold with [sharded repodata](https://github.com/conda/ceps/blob/main/cep-0016.md).
+
+**It would require extending the repodata schema, currently not formally standardized.**
+
+We propose to add a `run_export` field to each record that mimics the specification from CEP-12.
+
+If the `run_export` field is not present in the record it means no `run_export` information is stored with the record, and a fallback mechanism should be used to acquire the run-export information.
```

Comment:
Replicating the `run_export` field in the repodata would result in a lot of `"run_exports": {}` entries. These will compress well but would be it more efficient to store a top level field that indicates that records have an empty `run_exports` unless they are explicitly declared? This would add complexity at the benefit of smaller(?) repodata.

Reply:
We think that for shards, using `msgpack + zst`, it's fine.

Reply:
Plus the shards have much improved caching behavior (content-addressed) so that total download will always be much much lower vs. the current situation.